### PR TITLE
plugin: add GlobalDataProcess Plugin

### PR DESCRIPTION
**Pull request by WenheLI**
This PR adds a new type of plugin that allows users to process the whole dataset.

### CODE REVIEW

File: packages/cli/assets/pluginPackage/src/dataProcess.ts [hunk 1]
```diff
@@ -1,7 +1,6 @@
-import { DataProcessType, ArgsType, UniDataset } from '@pipcook/pipcook-core'
+import { DataProcessType, ArgsType, Sample } from '@pipcook/pipcook-core'
 
-const templateDataProcess: DataProcessType = async (data: UniDataset, args: ArgsType): Promise<UniDataset> => {
-  return {} as UniDataset;
+const templateDataProcess: DataProcessType = async (data: Sample, args: ArgsType): Promise<void> => {
```
Comment (rickycao-qy) [inline packages/cli/assets/pluginPackage/src/dataProcess.ts:3]: Does the local data-process type still have metadata for the second parameter? Cuz it still has the responsibility to change metadata as discussed

**Comment by yorkie** [comment]
The GlobalDataProcess is too long and implicit IMO, shall we have a better name for that?

**Comment by WenheLI** [comment]
@yorkie - How about `DatasetProcess` plugin?

**Comment by yorkie** [comment]
> @yorkie - How about `DatasetProcess` plugin?

I'm indefinitely +1 on this :p

File: packages/cli/assets/pluginPackage/src/globalDataProcess.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import { DatasetProcessType, ArgsType, UniDataset } from '@pipcook/pipcook-core'
+
+const templateDatasetProcess: DatasetProcessType = async (data: UniDataset, args: ArgsType): Promise<void> => {
+}
```
Comment (yorkie) [inline packages/cli/assets/pluginPackage/src/globalDataProcess.ts:4]:
```suggestion
};
```

File: packages/cli/assets/pluginPackage/src/globalDataProcess.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import { DatasetProcessType, ArgsType, UniDataset } from '@pipcook/pipcook-core'
+
+const templateDatasetProcess: DatasetProcessType = async (data: UniDataset, args: ArgsType): Promise<void> => {
```
Comment (yorkie) [inline packages/cli/assets/pluginPackage/src/globalDataProcess.ts:3]: Would you please add a comment line?

**Comment by yorkie** [comment]
@WenheLI we upgraded the Pipeline Schema in this PR, does that mean it's incompatible with the current db storage?

**Comment by WenheLI** [comment]
@yorkie - Nice catch! It seems `sequelize` does not provide auto-migrations. I think we need to implement [Migration](https://sequelize.org/master/manual/migrations.html) before merging this PR.

**Comment by yorkie** [comment]
It's related with #249, and I have a question about where to store the migration scripts? @WenheLI

**Comment by WenheLI** [comment]
@yorkie - Add a folder under `Daemon` called `scripts` or `migration` to store them might be appropriate?

**Comment by yorkie** [comment]
> @yorkie - Add a folder under `Daemon` called `scripts` or `migration` to store them might be appropriate?

Ok, then they will be executed when someone upgrades a newer daemon?

**Comment by WenheLI** [comment]
> > @yorkie - Add a folder under `Daemon` called `scripts` or `migration` to store them might be appropriate?
> 
> Ok, then they will be executed when someone upgrades a newer daemon?

Can we put a value like `DB_VERSION` in the database and `package.json`? If there is any updates regarding table structure, we manually update the version in `package.json`. And every time we start the daemon if the two versions are inconsistent, we run the script.

**Comment by yorkie** [comment]
Totally agreed, and we can remove [`specVersion`](https://github.com/alibaba/pipcook/blob/master/packages/daemon/src/model/job.ts#L47-L51) after it gets landed :)

**Comment by rickycao-qy** [comment]
LGTM once migration DB is ready

**Comment by rickycao-qy** [comment]
And please resolve confclits :)

**Comment by yorkie** [comment]
By the way, @WenheLI do we need to add a migration script for this PR?

**Comment by WenheLI** [comment]
> By the way, @WenheLI do we need to add a migration script for this PR?

The migration script is added by another PR(#397 )

**Pull request merged**